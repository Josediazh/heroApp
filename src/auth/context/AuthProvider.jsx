import { useEffect, useReducer } from "react"
import { AuthContext } from "./AuthContext"
import { AuthReducer } from "./AuthReducer"
import { types } from "../types/types"

const init = () => {

  const user = JSON.parse(localStorage.getItem('user'));

  return {
    logged: !!user,
    user: user
  }
}

export const AuthProvider = ({children}) => {

  const [state,dispatch] = useReducer(AuthReducer,{},init)  

  const onLogin = (name = '') => {

    const user = { id: 'ABC', name }

    const action ={
      type: types.login,
      payload: user
    }
    localStorage.setItem('user',JSON.stringify(user));
    dispatch(action);
  }

  const onLogout = () => {
    const action ={
      type: types.logout
    }

    localStorage.removeItem('user');
    dispatch(action);
  }
  

  return (
    <AuthContext.Provider value={{...state,login:onLogin,logout: onLogout}}>
        {children}
    </AuthContext.Provider >
  )
}