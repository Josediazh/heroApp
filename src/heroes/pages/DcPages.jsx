import { Navbar } from "../../ui/components/NavBar"
import { HeroesList } from "../components/HeroesList"


export const DcPages = () => {
  return (
    <>
      <Navbar />
      <h1>DC</h1>
      <HeroesList publisher={'DC Comics'}/>
    </>
  )
}