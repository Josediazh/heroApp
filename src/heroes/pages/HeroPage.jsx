import React, { useMemo } from 'react'
import { Navbar } from '../../ui/components/NavBar'
import { useNavigate, useParams } from 'react-router-dom'
import { GetByHeroId } from '../helpers/GetByHeroId';

export const HeroPage = () => {

  const {id} = useParams();
  const hero = useMemo(() => GetByHeroId(id),[id]);
  const heroImgPath = `/imgs/heroes/${hero.id}.jpg` 
  const charactersSplit = hero.characters.split(',').filter( character => character != hero.alter_ego );
  const navigate = useNavigate();

  const onBackNavigation = () => {
    navigate(-1);
  }


  if(!hero){
    return <Navigate to={'/'} />
  }

  return (
    <>
      <Navbar />
      <div className='container'>
        <div className='row mt-5'>
            <div className='col'>
              <img className='animate__animated animate__backInLeft img-thumbnail' src={heroImgPath} alt={hero.superhero} />
            </div>
            <div className='col'>
              <h1>{hero.superhero}</h1>
              <small className='text-muted'>{hero.publisher}</small>
              <hr />
              <p>{hero.alter_ego}</p>
              <p>{hero.first_appearance}</p>
              <p>{charactersSplit}</p>
              <button onClick={onBackNavigation} className='btn btn-primary'>Back</button>
            </div>
        </div>
      </div>
    </>
  )
}
