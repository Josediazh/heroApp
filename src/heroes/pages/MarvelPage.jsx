import { Navbar } from "../../ui/components/NavBar"
import { HeroesList } from "../components/HeroesList"

export const MarvelPage = () => {
  return (
    <>
      <Navbar />
      <h1>MarvelPage</h1>
      <HeroesList publisher={'Marvel Comics'}/>
    </>
  )
}