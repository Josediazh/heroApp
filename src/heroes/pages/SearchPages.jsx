import { Navbar } from '../../ui/components/NavBar'
import { useForm } from '../../hooks/useForm'
import { useLocation, useNavigate } from 'react-router-dom'
import queryString from 'query-string'
import { heroes } from '../../assets/data/HeroesData'
import { GetByHeroSearch } from '../helpers/GetByHeroSearch'
import { HeroesItem } from '../components/HeroesItem'

export const SearchPages = () => {

  const navigate = useNavigate();
  const location = useLocation();
  const {q = ''} = queryString.parse(location.search);
  const resultHeroes = GetByHeroSearch(q);
  const showSearch = (q.length === 0);
  const showError  = (q.length > 0) && resultHeroes.length === 0;

  const {formState, onInputChange, onResetForm, searchHero} = useForm({
    searchHero: q
  })

  const onSearchSubmit = (event) => {
    event.preventDefault();
    if (searchHero.trim().length <= 1) return;
    navigate(`?q=${searchHero}`);
  }

  return (
    <>
      <Navbar />
      <h1>Search Hero</h1>
      <hr />
      <div className="row p-5">
        <div className="col-4">
          <h4>Searching</h4>
          <hr />
          <form aria-label='form_search'  onSubmit={onSearchSubmit}>
            <input
            aria-label='input_search' 
            name="searchHero"
            placeholder="Search a hero"
            type="text" 
            value={searchHero}
            className="form-control"
            onChange={onInputChange}
            autoComplete="on"
            list="heroes"
            />
            <datalist id="heroes">
                {
                  heroes.map(({superhero}) => {
                    return <option key={superhero} value={superhero}>{superhero}</option>
                  })
                }
            </datalist>
            <button className="btn btn-outline-primary mt-2">Search</button>
          </form>
        </div>
        <div className="col-8">
          <h4 aria-label='results_number'>Results: {resultHeroes.length}</h4>
          <hr />
          <div className="alert alert-primary animate__animated animate__fadeIn" 
            style={{ display: showSearch ? '' : 'none' }}>
              Search a hero
          </div>

          <div aria-label='alert_error' className="alert alert-danger animate__animated animate__fadeIn" 
            style={{ display: showError ? '' : 'none' }}>
              No hero with <b>{ q }</b>
          </div>
          {
            resultHeroes.map( ({id,superhero,publisher,alter_ego,first_appearance,characters}) => {
              return (
                  <HeroesItem key={id} 
                  id={id}
                  superhero={superhero} 
                  alter_ego={alter_ego} 
                  first_appearance={first_appearance}
                  characters={characters}
                   />
              )
            })
          }
        </div>
      </div>
    </>
  )
}