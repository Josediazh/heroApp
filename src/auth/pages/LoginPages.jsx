import { useContext } from "react";
import { useNavigate } from "react-router-dom"
import { AuthContext } from "../context/AuthContext";


export const LoginPages = () => {

  const navigate = useNavigate();
  const {login} = useContext(AuthContext);
  const nameUser = 'Jose Luis'; 

  const onLoginUser = () => {

    const lastPath = localStorage.getItem('lastpath') || '/';
    login(nameUser);

    navigate(lastPath,{
      replace: true
    })

  }

  return (
    <>
      <div className="container mt-5">
        <h1>Login</h1>
        <hr />
        <button onClick={onLoginUser} className="btn btn-primary">Login</button>
      </div>
    </>
  )
}