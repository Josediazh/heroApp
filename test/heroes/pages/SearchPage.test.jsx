import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, useNavigate } from "react-router-dom";
import { SearchPages } from "../../../src/heroes/pages/SearchPages";
import { AuthProvider } from "../../../src/auth/context/AuthProvider";
import { AuthContext } from "../../../src/auth/context/AuthContext";

const mockUsedNavigate = jest.fn();

jest.mock('react-router-dom',() => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUsedNavigate
}))

describe('Prueba en el searchpage', () => {

    beforeEach( () => jest.clearAllMocks );

    const contextValue = {
        logged: true,
        user:{
            id: "123",
            name: "jose luis"
        }
      }

    test('debe de mostrarse con valores con defecto', () => { 

        const container = render(
            <MemoryRouter>
                <AuthContext.Provider value={contextValue}>
                    <SearchPages />
                </AuthContext.Provider>
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();

     })


     test('debe de mostrarse los resultados de la busqueda', () => { 

        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <AuthContext.Provider value={contextValue}>
                    <SearchPages />
                </AuthContext.Provider>
            </MemoryRouter>
        );

        expect(screen.getByLabelText('results_number').innerHTML).toBe('Results: 1');
        expect(screen.getByLabelText('hero_name').innerHTML).toBe('Batman')

     })

     test('debe de mostrar el mensaje de error', () => { 

        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <AuthContext.Provider value={contextValue}>
                    <SearchPages />
                </AuthContext.Provider>
            </MemoryRouter>
        );
        
        expect(screen.getByLabelText('alert_error').style.display).not.toBe('none');

     })


     test('debe de hacer el submit al formulario', () => { 

        render(
            <MemoryRouter initialEntries={['/search']}>
                <AuthContext.Provider value={contextValue}>
                    <SearchPages />
                </AuthContext.Provider>
            </MemoryRouter>
        );

        const inpuElement = screen.getByLabelText("input_search");
        const formElement = screen.getByLabelText("form_search");

        fireEvent.change(inpuElement, {target: {value: 'Superman'}});
        fireEvent.submit(formElement);

        expect(mockUsedNavigate).toHaveBeenCalled();
        expect(mockUsedNavigate).toHaveBeenCalledWith('?q=Superman');

     })
});