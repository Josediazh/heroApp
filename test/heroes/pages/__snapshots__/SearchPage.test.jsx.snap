// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en el searchpage debe de mostrarse con valores con defecto 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="navbar navbar-expand-sm navbar-dark bg-dark p-2"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          Asociaciones
        </a>
        <div
          class="navbar-collapse"
        >
          <div
            class="navbar-nav"
          >
            <a
              class="nav-item nav-link "
              href="/marvel"
            >
              Marvel
            </a>
            <a
              class="nav-item nav-link "
              href="/dc"
            >
              DC
            </a>
            <a
              class="nav-item nav-link "
              href="/search"
            >
              <i
                class="bi bi-search"
              />
            </a>
          </div>
        </div>
        <div
          class="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end"
        >
          <ul
            class="navbar-nav ml-auto"
          >
            <span
              aria-label="username"
              class="nav-item nav-link text-info"
            >
              jose luis
            </span>
            <button
              class="nav-item nav-link info btn"
            >
              Logout
            </button>
          </ul>
        </div>
      </nav>
      <h1>
        Search Hero
      </h1>
      <hr />
      <div
        class="row p-5"
      >
        <div
          class="col-4"
        >
          <h4>
            Searching
          </h4>
          <hr />
          <form
            aria-label="form_search"
          >
            <input
              aria-label="input_search"
              autocomplete="on"
              class="form-control"
              list="heroes"
              name="searchHero"
              placeholder="Search a hero"
              type="text"
              value=""
            />
            <datalist
              id="heroes"
            >
              <option
                value="Batman"
              >
                Batman
              </option>
              <option
                value="Superman"
              >
                Superman
              </option>
              <option
                value="Flash"
              >
                Flash
              </option>
              <option
                value="Green Lantern"
              >
                Green Lantern
              </option>
              <option
                value="Green Arrow"
              >
                Green Arrow
              </option>
              <option
                value="Wonder Woman"
              >
                Wonder Woman
              </option>
              <option
                value="Martian Manhunter"
              >
                Martian Manhunter
              </option>
              <option
                value="Robin/Nightwing"
              >
                Robin/Nightwing
              </option>
              <option
                value="Blue Beetle"
              >
                Blue Beetle
              </option>
              <option
                value="Black Canary"
              >
                Black Canary
              </option>
              <option
                value="Spider Man"
              >
                Spider Man
              </option>
              <option
                value="Captain America"
              >
                Captain America
              </option>
              <option
                value="Iron Man"
              >
                Iron Man
              </option>
              <option
                value="Thor"
              >
                Thor
              </option>
              <option
                value="Hulk"
              >
                Hulk
              </option>
              <option
                value="Wolverine"
              >
                Wolverine
              </option>
              <option
                value="Daredevil"
              >
                Daredevil
              </option>
              <option
                value="Hawkeye"
              >
                Hawkeye
              </option>
              <option
                value="Cyclops"
              >
                Cyclops
              </option>
              <option
                value="Silver Surfer"
              >
                Silver Surfer
              </option>
            </datalist>
            <button
              class="btn btn-outline-primary mt-2"
            >
              Search
            </button>
          </form>
        </div>
        <div
          class="col-8"
        >
          <h4
            aria-label="results_number"
          >
            Results: 
            0
          </h4>
          <hr />
          <div
            class="alert alert-primary animate__animated animate__fadeIn"
          >
            Search a hero
          </div>
          <div
            aria-label="alert_error"
            class="alert alert-danger animate__animated animate__fadeIn"
            style="display: none;"
          >
            No hero with 
            <b />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar navbar-expand-sm navbar-dark bg-dark p-2"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Asociaciones
      </a>
      <div
        class="navbar-collapse"
      >
        <div
          class="navbar-nav"
        >
          <a
            class="nav-item nav-link "
            href="/marvel"
          >
            Marvel
          </a>
          <a
            class="nav-item nav-link "
            href="/dc"
          >
            DC
          </a>
          <a
            class="nav-item nav-link "
            href="/search"
          >
            <i
              class="bi bi-search"
            />
          </a>
        </div>
      </div>
      <div
        class="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end"
      >
        <ul
          class="navbar-nav ml-auto"
        >
          <span
            aria-label="username"
            class="nav-item nav-link text-info"
          >
            jose luis
          </span>
          <button
            class="nav-item nav-link info btn"
          >
            Logout
          </button>
        </ul>
      </div>
    </nav>
    <h1>
      Search Hero
    </h1>
    <hr />
    <div
      class="row p-5"
    >
      <div
        class="col-4"
      >
        <h4>
          Searching
        </h4>
        <hr />
        <form
          aria-label="form_search"
        >
          <input
            aria-label="input_search"
            autocomplete="on"
            class="form-control"
            list="heroes"
            name="searchHero"
            placeholder="Search a hero"
            type="text"
            value=""
          />
          <datalist
            id="heroes"
          >
            <option
              value="Batman"
            >
              Batman
            </option>
            <option
              value="Superman"
            >
              Superman
            </option>
            <option
              value="Flash"
            >
              Flash
            </option>
            <option
              value="Green Lantern"
            >
              Green Lantern
            </option>
            <option
              value="Green Arrow"
            >
              Green Arrow
            </option>
            <option
              value="Wonder Woman"
            >
              Wonder Woman
            </option>
            <option
              value="Martian Manhunter"
            >
              Martian Manhunter
            </option>
            <option
              value="Robin/Nightwing"
            >
              Robin/Nightwing
            </option>
            <option
              value="Blue Beetle"
            >
              Blue Beetle
            </option>
            <option
              value="Black Canary"
            >
              Black Canary
            </option>
            <option
              value="Spider Man"
            >
              Spider Man
            </option>
            <option
              value="Captain America"
            >
              Captain America
            </option>
            <option
              value="Iron Man"
            >
              Iron Man
            </option>
            <option
              value="Thor"
            >
              Thor
            </option>
            <option
              value="Hulk"
            >
              Hulk
            </option>
            <option
              value="Wolverine"
            >
              Wolverine
            </option>
            <option
              value="Daredevil"
            >
              Daredevil
            </option>
            <option
              value="Hawkeye"
            >
              Hawkeye
            </option>
            <option
              value="Cyclops"
            >
              Cyclops
            </option>
            <option
              value="Silver Surfer"
            >
              Silver Surfer
            </option>
          </datalist>
          <button
            class="btn btn-outline-primary mt-2"
          >
            Search
          </button>
        </form>
      </div>
      <div
        class="col-8"
      >
        <h4
          aria-label="results_number"
        >
          Results: 
          0
        </h4>
        <hr />
        <div
          class="alert alert-primary animate__animated animate__fadeIn"
        >
          Search a hero
        </div>
        <div
          aria-label="alert_error"
          class="alert alert-danger animate__animated animate__fadeIn"
          style="display: none;"
        >
          No hero with 
          <b />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
